src/client/api/rpc.c

```diff
@@ -161,8 +164,16 @@ daos_rpc_proto_query(crt_opcode_t base_opc, uint32_t *ver_array, int count, int
 		D_GOTO(out_detach, rc = -DER_NOMEM);
 
 	/** select a random rank to issue the proto query rpc to */
-	rproto->nr_ranks   = dc_mgmt_net_get_num_srv_ranks();
-	rproto->ep.ep_rank = get_rand_rank();
+	nr_ranks = dc_mgmt_net_get_num_srv_ranks();
+	if (nr_ranks == 0) {
+		D_ERROR("failed to select an attached ranks: no attached ranks");
+		D_GOTO(out_free, -DER_NONEXIST);
+	}
+	rproto->rank_idx = d_rand() % nr_ranks;
+	rank             = dc_mgmt_net_get_srv_rank(rproto->rank_idx);
+	D_ASSERT(rank != CRT_NO_RANK);
```

Comment:
is there any chance this can be true in a situation where the agent is not running or so. asserts always scare me.

Reply:
It should never happen if `nr_ranks` is > 0 as the `nr_ranks` is equal to the number of elements in the conversion table used by the `dc_mgmt_net_get_srv_rank()` function.

Reply:
The idea of this ASSERT and the one from line 126 is to detect some future regression which makes the previous properties not any more correct (i.e. for example concurrent access to the global variables `g_num_serv_ranks` and `g_serv_ranks`).